cpp/src/traversal/bfs.cu

```diff
@@ -474,10 +476,14 @@ void bfs(experimental::GraphCSRView<VT, ET, WT> const &graph,
          const VT start_vertex,
          bool directed)
 {
-  CUGRAPH_EXPECTS(typeid(VT) == typeid(int), "Unsupported vertex id data type, please use int");
-  CUGRAPH_EXPECTS(typeid(ET) == typeid(int), "Unsupported edge id data type, please use int");
-  CUGRAPH_EXPECTS((typeid(WT) == typeid(float)) || (typeid(WT) == typeid(double)),
-                  "Unsupported weight data type, please use float or double");
+  static_assert(std::is_same<VT, unsigned>::value || std::is_same<VT, int>::value ||
+                  std::is_same<VT, long>::value,
+                "Unsupported vertex id data type, please use unsigned, int, or long");
+
+  static_assert(std::is_same<VT, ET>::value, "VT and ET should be the same time");
+
+  static_assert(std::is_same<WT, float>::value || std::is_same<WT, double>::value,
+                "Unsupported weight data type, please use float or double");
```

Comment:
These are more dependent on explicit instantiation rather than fundamental requirements of the algorithm and these will be detected in link time anyway.

Better be

```
static_assert(std::is_integral<VT>::value, "...");
static_assert(std::is_integral<ET>::value, "...");
static_assert(std::is_floating_point<WT>::value, "...");  // actually, this is unnecessary for BFS
```

Reply:
My concern was that is_integral is too wide (it includes all widths including 8 and 16), and unless the kernel is really tested to work for all widths, it will likely break if you instantiate it with anything other than what is tested. There are implicit assumptions in places for sizes to be at least 32 bits. One example is the dimx params for block/grid etc. So this assert was to signal the current state of affairs. Same for std::is_same<VT, ET>::value.

Reply:
OK, I see... If you expect this code to break if sizeof(ET) > sizeof(VT), then, better avoid relaxing the check.

Reply:
I'll update the check to assert is_integral, size at least 4 bytes, and VT and ET are the same type. I think the kernels need one additional pass to make them really type agnostic and safe. Need to go through all the arithmetic line by line to check for breakage. There may also be orthogonal overflow issues. One way to weed them out would be to use graphs of sizes at the limit of the type. For eg., things like ceil division (a + K -1)/K will start breaking at that point.